Adding new groovy scripts for httpd, openjdk and tomcat

vaibhavsood opened the pull request:
Hi Tianon,

I have created groovy scripts for httpd, openjdk and tomcat (these are missing in this repo but have official-image docker files). I have tested them on my local jenkins setup with ppc64le hardware and wanted to contribute them to the multiarch repo

For openjdk, i needed to add a case: statement for ppc64le since openjdk9 on ppc has a missing dependent package (udev), httpd and tomcat work as-is

Please do let me know if there are any changes/improvements i can make to the scripts

Thanks!
Vaibhav

Commit:
Adding new groovy scripts for httpd, openjdk(ppc64le specific) and tomcat

Commit:
Update docs.groovy and views.groovy for the new groovy scripts

tianon commented on dsl/docker-multiarch/images/openjdk.groovy:
```diff
@@ -0,0 +1,44 @@
+import vars.multiarch
```
I'm really torn on whether to call this file `openjdk.groovy` or `java.groovy`.  This entire repo was misnamed to begin with and I'm really pretty sad that we don't have a clean way to get around the fallout it created.  Regardless, we do need to make sure this image gets built, tagged, and pushed as `PREFIX/java` _not_ `PREFIX/openjdk` (since we need to be as close to faithful to the official images as we can be) -- I'm OK with `openjdk.groovy` if we also add the override in `vars/multiarch.groovy` in `def static image(clsName)` to ensure it tags, builds, and pushes as `java`.

vaibhavsood replied:
Fixed this. Added the override 'openjdk': 'java' to vars/multiarch.groovy and tested, image now builds as 'java'

tianon commented on dsl/docker-multiarch/images/openjdk.groovy:
```diff
@@ -0,0 +1,44 @@
+			}
+		}
+		triggers {
+			upstream("docker-${arch}-debian, docker-${arch}-buildpack", 'UNSTABLE')
```
This should only be `docker-${arch}-buildpack`: (no `-debian`)

``` console
$ grep -n FROM */Dockerfile
openjdk-6-jdk/Dockerfile:7:FROM buildpack-deps:wheezy-scm
openjdk-6-jre/Dockerfile:7:FROM buildpack-deps:wheezy-curl
openjdk-7-jdk/Dockerfile:7:FROM buildpack-deps:jessie-scm
openjdk-7-jre/Dockerfile:7:FROM buildpack-deps:jessie-curl
openjdk-8-jdk/Dockerfile:7:FROM buildpack-deps:jessie-scm
openjdk-8-jre/Dockerfile:7:FROM buildpack-deps:jessie-curl
openjdk-9-jdk/Dockerfile:7:FROM buildpack-deps:sid-scm
openjdk-9-jre/Dockerfile:7:FROM buildpack-deps:sid-curl
```

vaibhavsood replied:
Removed the unnecessary dependent

tianon commented on dsl/docker-multiarch/images/openjdk.groovy:
```diff
@@ -0,0 +1,44 @@
+			shell(multiarch.templateArgs(meta, ['dpkgArch']) + '''
+case "$dpkgArch" in
+	ppc64el)
+		sed -i 's!amd64!ppc64el!g' */Dockerfile
```
Couldn't this bit be generalized to something like `sed -i "s!amd64!$dpkgArch!g" */Dockerfile` (and thus moved out of the case)?  Wouldn't it need to be in order to support arches that aren't `ppc64le`?  (We support `i386` which should build fine on an `amd64` box, so you could add a 32bit node to your Jenkins to test these on too, which might help in this regard.)

vaibhavsood replied:
Changed to the generalized line as suggested, tested on ppc64le, works correctly

tianon commented on dsl/docker-multiarch/images/openjdk.groovy:
```diff
@@ -0,0 +1,44 @@
+case "$dpkgArch" in
+	ppc64el)
+		sed -i 's!amd64!ppc64el!g' */Dockerfile
+		sed -i 's!unzip!unzip udev!' openjdk-9-jdk/Dockerfile
```
This one has me worried -- you mentioned it was working around some issue with the package installing properly?  Is there already a Debian bug filed for it?

vaibhavsood replied:
I have removed these lines as I no longer see this error for openjdk9 for the subversion 9~b102. However now the latest subversion has been changed to 9~b107 in this commit:

https://github.com/docker-library/openjdk/commit/d7f632dcc518cf71752e95387d05cbb7f9156e1d

 but this version is not available yet for ppc64le so testing give package unavailable error for openjdk9

Not sure how to handle this but removing these lines for now as they are unnecessary anyways in this groovy script

tianon commented on dsl/docker-multiarch/images/openjdk.groovy:
```diff
@@ -0,0 +1,44 @@
+		sed -i 's!unzip!unzip udev!' openjdk-9-jdk/Dockerfile
+esac
+
+sed -i "s!^FROM !FROM $prefix/!" */Dockerfile
```
I'd prefer to move this line to the top of the `script` block here just so we have symmetry with other files here (https://github.com/tianon/jenkins-groovy/blob/b11361e40e555e23f0d64be118b25433ca1d18b7/dsl/docker-multiarch/images/gcc.groovy#L23-L25 for example).

vaibhavsood replied:
Restored this back to the top

tianon commented on dsl/docker-multiarch/views.groovy:
```diff
@@ -10,7 +10,9 @@ def images = [
 	'hello',
+	'httpd',
 	'irssi',
+	'java',
```
Since this specific list is filenames, not image names, this should be `openjdk`.

vaibhavsood replied:
Corrected this

vaibhavsood commented:
Hi Tianon,

Sorry for the slightly delayed response, I have updated the diffs addressing your review comments (thanks for those!), my comments are inline. 

Please do let me know of any further comments/suggestions

Thanks!
Vaibhav

tianon commented:
Nice, no worries!  I obviously understand that we all need to be a little asynchronous sometimes. :innocent:

LGTM! :+1:

vaibhavsood commented:
Sure thanks! :smile:
